migrations/1686291754032-copy-legacy-tools-to-ctl.js

```diff
@@ -0,0 +1,374 @@
+							type: String,
+							enum: ['global', 'school', 'context'],
+							required: true,
+						},
+						location: {
+							type: String,
+							enum: ['path', 'body', 'query'],
+							required: true,
+						},
+						type: {
+							type: String,
+							enum: ['string', 'number', 'boolean', 'auto_contextid', 'auto_contextname', 'auto_schoolid'],
+							required: true,
+						},
+						isOptional: { type: Boolean, required: true },
+					},
+				},
+			],
+
+			isHidden: { type: Boolean, required: true },
+			openNewTab: { type: Boolean, required: true },
+			version: { type: Number, required: true },
+			createdAt: { type: Date, default: Date.now },
+			updatedAt: { type: Date, default: Date.now },
+		},
+		{
+			timestamps: true,
+		}
+	),
+	'external_tools'
+);
+
+const SchoolExternalTool = mongoose.model(
+	'school_external_tool0906202311483',
+	new mongoose.Schema(
+		{
+			_id: { type: Schema.Types.ObjectId, required: true },
+			tool: { type: Schema.Types.ObjectId, ref: 'externalTool', required: true },
+			school: { type: Schema.Types.ObjectId, required: true },
+			schoolParameters: [{ type: { key: { type: String }, value: { type: String } } }],
+			toolVersion: { type: Number, required: true },
+			createdAt: { type: Date, default: Date.now },
+			updatedAt: { type: Date, default: Date.now },
+		},
+		{
+			timestamps: true,
+		}
+	),
+	'school_external_tools'
+);
+
+const ContextExternalTool = mongoose.model(
+	'context_external_tool0906202311484',
+	new mongoose.Schema(
+		{
+			_id: { type: Schema.Types.ObjectId, required: true },
+			schoolTool: { type: Schema.Types.ObjectId, ref: 'schoolExternalTool', required: true },
+			contextId: { type: Schema.Types.ObjectId, required: true },
+			contextType: { type: String, enum: ['course'], required: true },
+			contextToolName: { type: String },
+			parameters: [{ type: { key: { type: String }, value: { type: String } } }],
+			toolVersion: { type: Number, required: true },
+			createdAt: { type: Date, default: Date.now },
+			updatedAt: { type: Date, default: Date.now },
+		},
+		{
+			timestamps: true,
+		}
+	),
+	'context_external_tools'
+);
+
+const Course = mongoose.model(
+	'course0906202311485',
+	new mongoose.Schema(
+		{
+			_id: { type: Schema.Types.ObjectId, required: true },
+			school: { type: Schema.Types.ObjectId, required: true },
+		},
+		{
+			timestamps: true,
+		}
+	),
+	'courses'
+);
+
+function mapToExternalToolParameter(ltiToolTemplate) {
+	return ltiToolTemplate.customs.map((parameter) => {
+		return {
+			name: parameter.key,
+			displayName: parameter.key,
+			description: '',
+			default: '',
+			regex: '',
+			regexComment: '',
+			scope: 'context',
+			location: 'body',
+			type: 'string',
+			isOptional: true,
+		};
+	});
+}
+
+function mapToCustomParameterEntry(externalToolParameters, ltiToolCustomes) {
+	return externalToolParameters.map((parameter, index) => {
+		return {
+			name: parameter.name,
+			value: ltiToolCustomes[index].value,
```

Comment:
you should not use the index, since both array do not require the indexes to be the same e.g. external parameters are ['course1Param', 'course2Param'] and your lti tool in course 2 only has ['course2Param'] then accessing index 0 would write 'course2Param' to 'course1Param' and index 1 would throw you an error since `undefined.value` is not a field.